Advent25 - Sara Joy - Discover Dialog (#227)
matuzo/HTMHell: hell/adventcalendar/2025/9/index.md

```html
<form method="dialog">
  <button>Close</button>
</form>
```

Dialogs can also be closed without using this form method, instead by activating some JS (maybe also with a button):

```js
document.querySelector("dialog").close();
// or
document.querySelector("dialog").requestClose();
```

This [`requestClose()`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDialogElement/requestClose) is a newly available addition, allowing access to the [cancel event](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDialogElement/cancel_event), otherwise thrown before the [close event](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDialogElement/close_event) when exiting a modal with the `Esc` key. This may be useful to create _"are you sure?"_ type interactions. (Please do so with care if you must! We'll have no deceptive patterns here..)

> ### Light dimissal

Reviewer: ```suggestion
### Light dismissal
```

- [Getting Creative With HTML Dialog by Andy Clarke at css-tricks.com](https://css-tricks.com/getting-creative-with-html-dialog/)
- [Animating the Dialog Element by Matthew Morete at frontendmasters.com](https://frontendmasters.com/blog/animating-dialog/)

## Scrollin', scrollin', scrollin'...

That by default the website under your modal is still scrollable may or may not bother you. It bothers _me_!

There are ways and means to stop that happening, thankfully.

A classic and elegant way made possible with the sort-of-new-but-well-supported [`:has()`](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Selectors/:has) pseudo-class and [`scrollbar-gutter`](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Properties/scrollbar-gutter) property is the following:

```css
/* Check whether any dialogs are open */
html:has(dialog[open]) {
  /* Poof! No more scrolling! */
  overflow: hidden;
  /* keep the scrollbar width */
  scrollbar-gutter: stable;
}
```

The `scrollbar-gutter` property keeps the text on the page from reflowing with a jerk, when it suddenly becomes one scrollbar-width wider as the rest of the page beyond the viewport is hidden.

> <p class="highlight">💡

Reviewer: Why not include it in your example right away?

```html
html:has(dialog[open]:modal) {
  /* Poof! No more scrolling! */
  overflow: hidden;
  /* keep the scrollbar width */
  scrollbar-gutter: stable;
}
```

I heartily recommend reading [Where to Put Focus When Opening a Modal Dialog](https://adrianroselli.com/2025/06/where-to-put-focus-when-opening-a-modal-dialog.html?Theme=Light) by Adrian Roselli to help you decide where the focus will land on your dialog. Then read [O dialog focus, where art thou?](https://www.matuzo.at/blog/2023/focus-dialog/) to help you consistently get it where you want it.

<p class="highlight">
  If you read the MDN page on dialogs and are bothered by <a href="https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/dialog#attributes">this warning about <code>tabindex</code> on <code>&lt;dialog&gt;</code></a>, you may either take Adrian or Manuel's word that it's OK, or you can enjoy this deep dive into <a href="https://www.matuzo.at/blog/2025/whats-an-interactive-element">What's an interactive element?</a>. Enjoy!
</p>

## Dialogs discovered

> I hope this has been enough to convince you that its worth using dialogs, if you aren't already, or maybe you've found a couple of interesting tidbits you otherwise werent aware of.

Reviewer: ```suggestion
I hope this has been enough to convince you that its worth using dialogs, if you aren't already, or maybe you've found a couple of interesting tidbits you otherwise weren't aware of.
```

```html
<dialog id="meetDialog">
  <p>
    Hello 👋
    <br />Nice to meet you!
  </p>
  <form method="dialog">
    <button>Close</button>
  </form>
</dialog>
```

## Modal and non-modal

We're probably more familiar with the modal variety of dialog, intended to (hopefully only briefly) stop you interacting with the page under it. But there are also plenty of non-modal dialogs to be seen around the web, for example <button id="openToastDialog">toast</button> notifications.

<dialog
  id="toastDialog"
  closedby="any"
  style="
    margin: 0;
    border-width: 0.25em 1em 1em 0.25em;
    border-style: solid;
    border-radius: 1.5em 2em 0.5em 1em;
    border-color: saddlebrown;
    background-color: burlywood;
    box-shadow: 1em 1em 1.5em -0.5em #8888;
    rotate: 10deg;
    position: fixed;
    left: 3em;
    bottom: 3em;
    z-index: 10;
  "
>
  <p>
    Pop! Do I look
    <br> a little burned?
  </p>
  <form method="dialog">
    <button>Close</button>
  </form>
</dialog>

> Dialogs can give us both modal and non-modal variants, depending on how we open them. In the case of more than one dialog being opened (assuming no close action happening on the previous dialogs), the most recently opened dialog will appear on top.

Reviewer 2: would say dialog element for avoidance of confusion with the dictionary word 'dialog', and “gives” instead of “can give”: 

```suggestion
The `dialog` element gives us both modal and non-modal variants, depending on how we open them. In the case of more than one dialog being opened (assuming no close action happening on the previous dialogs), the most recently opened dialog will appear on top.
```

## Opening with JS

To open a dialog, we use one of two JavaScript functions on a dialog element:

```js
.show()
.showModal()
```

Likely based on the click of a button, you will run something like the following (you will likely use a much more specific `querySelector`):

```js
document.querySelector("dialog").show();
// or
document.querySelector("dialog").showModal();
```

<p class="highlight">💡
  There is a declarative method available to <a href="">invoke</a> dialogs, but for now I will stick to what is currently available across the main browsers.
</p>

> In the CodePen below, both kinds of dialog opening method are available. You'll notice that you can still interact with the Open Modal button after having opened the dialog, but not vice-versa. The modal can be closed with the Esc key, the normal modal cannot.

Reviewer 2: I think with “normal modal” you mean the non modal dialog? Maybe say “non-modal dialog” instead of “normal modal”?

```html
<form method="dialog">
  <button>Close</button>
</form>
```

Dialogs can also be closed without using this form method, instead by activating some JS (maybe also with a button):

```js
document.querySelector("dialog").close();
// or
document.querySelector("dialog").requestClose();
```

This [`requestClose()`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDialogElement/requestClose) is a newly available addition, allowing access to the [cancel event](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDialogElement/cancel_event), otherwise thrown before the [close event](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDialogElement/close_event) when exiting a modal with the `Esc` key. This may be useful to create _"are you sure?"_ type interactions. (Please do so with care if you must! We'll have no deceptive patterns here..)

### Light dismissal

> You're probably already trying to close the dialogs by clicking away, which is a common pattern and is often referred to as _light dismiss_.

Reviewer 2: maybe say “clicking outside” instead of “clicking away”? I think that's what you mean but is more specific? 

Would also suggest linking “light dismiss” to this location in the spec for people who want to know the nitty gritty? https://html.spec.whatwg.org/dev/interactive-elements.html#dialog-light-dismiss

      <span>See the Pen <a href="https://codepen.io/sarajw/pen/yyepzaX">
  Dialogs with light dismiss (JS)</a> by Sara (<a href="https://codepen.io/sarajw">@sarajw</a>)
  on <a href="https://codepen.io">CodePen</a>.</span>
      </p>
      <script async src="https://public.codepenassets.com/embed/index.js"></script>

<p class="highlight">💡
  You may have spotted we selected <code>dialog[open]</code> in the JS above.
  <br>
  <br>Yes, open dialogs do get <code>open</code> added as an attribute, and it is <em>possible</em> to toggle this to open and close a dialog, but it is <a href="https://developer.mozilla.org/en-US/docs/Web/API/HTMLDialogElement/open">not recommended</a>. It will always open as a non-modal dialog when done in this way, and may lead to confused or missing <code>close</code> events.
</p>

I am selecting `:modal` dialogs only in the code and demonstration above - but you can omit this if you want the click-away to also close any _non_-modal dialogs that happen to be open at the same time.

> To achieve light dismiss on non-modal dialogs alone is trickier unfortunately, as they have no backdrop to play with...

Reviewer 2: `popover` exists for this!

You may have purposefully avoided it in this post but it is a cross browser supported API. You _could_ consider adding a link to [my post](https://hidde.blog/dialog-modal-popover-differences/) for the differences.

Author: Yeah I signalled you about this :)

I'm afraid I'm going to skip it, because I tried adding `popover` as an attribute to dialog and it didn't seem to magically add light dismiss - I tried a few other things and didn't really have luck making it happen simply, probably needing some other finicking to get it to work. As such, I'm not going to go there this time!

But there is some light at the end of the dismiss tunnel with the <a href="https://developer.mozilla.org/en-US/docs/Web/API/HTMLDialogElement/closedBy">`closedby`</a> property, which is very nearly available across the big browsers - so please feel free to poke any friendly Safari developers you might know! Then it could look as simple as the below:

<p class="codepen" data-height="300" data-default-tab="html,result" data-slug-hash="myVNLmb" data-pen-title="Dialogs with light dismiss (closedby)" data-editable="true" data-user="sarajw" style="height: 300px; box-sizing: border-box; display: flex; align-items: center; justify-content: center; border: 2px solid; margin: 1em 0; padding: 1em;">
    <span>See the Pen <a href="https://codepen.io/sarajw/pen/myVNLmb">
Dialogs with light dismiss (closedby)</a> by Sara (<a href="https://codepen.io/sarajw">@sarajw</a>)
on <a href="https://codepen.io">CodePen</a>.</span>
    </p>
    <script async src="https://public.codepenassets.com/embed/index.js"></script>

## (Some) styling built in

Apart from limiting the piratey _lorem ipsum_ column width and centering it, no other custom styling has occurred in the CodePen demos.

> By default, the normal dialog has no backdrop, while the modal dialog's `::backdrop` has a subtle transparent grey, and the dialog itself is centered vertically as well as horizontally on the page.

Reviewer 2: again, would say non-modal instead of normal

html:has(dialog[open]:modal) {
  /* Poof! No more scrolling! */
  overflow: hidden;
  /* keep the scrollbar width */
  scrollbar-gutter: stable;
}
```

The `scrollbar-gutter` property keeps the text on the page from reflowing with a jerk, when it suddenly becomes one scrollbar-width wider as the rest of the page beyond the viewport is hidden.

<p class="highlight">💡
  If you have both non-modal and modal dialogs on your page, be sure to use <code>:modal</code> in your selector so this only triggers on open <em>modal</em> dialogs.
</p>

<p class="codepen" data-height="300" data-default-tab="css,result" data-slug-hash="OPMzjBY" data-pen-title="Modal dialog with scroll-stop" data-editable="true" data-user="sarajw" style="height: 300px; box-sizing: border-box; display: flex; align-items: center; justify-content: center; border: 2px solid; margin: 1em 0; padding: 1em;">
      <span>See the Pen <a href="https://codepen.io/sarajw/pen/OPMzjBY">
  Modal dialog with scroll-stop</a> by Sara (<a href="https://codepen.io/sarajw">@sarajw</a>)
  on <a href="https://codepen.io">CodePen</a>.</span>
      </p>
      <script async src="https://public.codepenassets.com/embed/index.js"></script>

> You've may have noticed - if you can see a scrollbar - that the scrollbar gutter isn't taking on the backdrop style. Yeah. Go ahead and comment out the `scrollbar-gutter: stable;` line in the CSS, open/close the modal a few times, and see whether the scrollbar appearing and disappearing bothers you more. It might not - but it bothers _me_!

Reviewer 2: typo: 

```suggestion
You may have noticed - if you can see a scrollbar - that the scrollbar gutter isn't taking on the backdrop style. Yeah. Go ahead and comment out the `scrollbar-gutter: stable;` line in the CSS, open/close the modal a few times, and see whether the scrollbar appearing and disappearing bothers you more. It might not - but it bothers _me_!
```

I heartily recommend reading [Where to Put Focus When Opening a Modal Dialog](https://adrianroselli.com/2025/06/where-to-put-focus-when-opening-a-modal-dialog.html?Theme=Light) by Adrian Roselli to help you decide where the focus will land on your dialog. Then read [O dialog focus, where art thou?](https://www.matuzo.at/blog/2023/focus-dialog/) to help you consistently get it where you want it.

<p class="highlight">
  If you read the MDN page on dialogs and are bothered by <a href="https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/dialog#attributes">this warning about <code>tabindex</code> on <code>&lt;dialog&gt;</code></a>, you may either take Adrian or Manuel's word that it's OK, or you can enjoy Manuel's deep dive into <a href="https://www.matuzo.at/blog/2025/whats-an-interactive-element">What's an interactive element?</a>. Enjoy!
</p>

> 

Reviewer: That chapter is misleading because it sounds like devs have to decide where to put focus. You should explain what happens by default and briefly mention situations where devs may want to put focus elsewhere.

Author: Hopefully the rewrite here works better:

> As it opens, by default the focus will land on either the whole dialog, or if it has one, the first focusable element (whether it's a close button or not). A sighted user with a pointing device will probably not tend to notice where the focus lands - but it's very important as a screen reader user, as the focus determines where in a document the reader lands and begins reading from.

> In case of needing the focus to land somewhere specific for your use case, I heartily recommend reading [Where to Put Focus When Opening a Modal Dialog](https://adrianroselli.com/2025/06/where-to-put-focus-when-opening-a-modal-dialog.html?Theme=Light) by Adrian Roselli to help you decide where the focus will land on your dialog. Then read [O dialog focus, where art thou?](https://www.matuzo.at/blog/2023/focus-dialog/) to help you consistently get it where you want it.
